### ISSUE TITLE
AuthCodeGrant $this->encrypt failure

### ISSUE THREAD

**Issue by aluciffer** [thephpleague/oauth2-server#890]
Hello.

I have a weird behavior, where this->encrypt has a different output in AuthCodeGrant [around line 365](https://github.com/thephpleague/oauth2-server/blob/2375b8c7f3bd757bfa9a9f8d5a13e51d9fc460c4/src/Grant/AuthCodeGrant.php#L358).
These following two different calls output different encrypted strings, which in turn, in my installation lead to a failure to correctly decrypt the autorization code.

```php
$params =[
    'code'  => $this->encrypt(
        json_encode(
            $payload
         )
    ),
    'state' => $authorizationRequest->getState(),
];

echo $this->makeRedirectUri(
    $finalRedirectUri,
    $params
) . PHP_EOL .
.$this->makeRedirectUri(
    $finalRedirectUri,
    [
        'code'  =>  $this->encrypt(
            json_encode(
                $payload
            )
         ),
        'state' => $authorizationRequest->getState(),
    ]
);
```

I presume it has something to do with the encrypt method Defuse\Crypto\Crypto->encrypt being statically defined and here being dynamically called?

I must add that i got here from an original laravel 5.6 installation trying to setup an oauth server, and testing it with a guzzlehttp client. Interestingly enough, if i work with a manual curl client, the code gets correctly passed back & forth.

**Comment by Sephster** [comment]
I know this is likely an obvious question but bear with me so we can rule out everything.

* Have you var dumped the `$payload` in both instances?
* Have you confirmed that the key being used for encryption (usually set in the `setEncryptionKey` function in the Crypt Trait) is identical for both calls?

**Comment by aluciffer** [comment]
Hi Sephster, no worries, i will bear with anything, because i've spent days on this topic and i'm completely puzzled.
I keep obtaining the two different outputs running literally the code above, how can $payload be different?, if i dump it out, where exactly should i dump it to see if it's different?
If i have the two lines one below the other in the code, how can the result be different?

To follow your suggestion, i have changed crypttrait as follows:
```
/**
     * Decrypt data with a public key.
     *
     * @param string $encryptedData
     *
     * @throws \LogicException
     *
     * @return string
     */
    protected function decrypt($encryptedData)
    {
        $fp = fopen('decrypt.trait.txt', 'a');
        fwrite($fp, 'call of '.$_SERVER['SERVER_NAME'] . ' ' . print_R($_REQUEST,1). PHP_EOL. ' with encKey='.print_r($this->encryptionKey,1).PHP_EOL.PHP_EOL);
        fclose($fp);

        try {
            return Crypto::decryptWithPassword($encryptedData, $this->encryptionKey);
        } catch (\Exception $e) {
            throw new \LogicException($e->getMessage());
        }
    }

    /**
     * Set the encryption key
     *
     * @param string $key
     */
    public function setEncryptionKey($key = null)
    {
        $fp = fopen('crypt.trait.txt', 'a');
        fwrite($fp, 'call of '.$_SERVER['SERVER_NAME'] . ' ' . print_R($_REQUEST,1). PHP_EOL. ' with key='.print_r($key,1).PHP_EOL.PHP_EOL);
        fclose($fp);
        $this->encryptionKey = $key;
    }
```

And in AuthCodeGrant.php, the function completeAuthorizationRequest as follows:
```
/**
     * {@inheritdoc}
     */
    public function completeAuthorizationRequest(AuthorizationRequest $authorizationRequest)
    {
        if ($authorizationRequest->getUser() instanceof UserEntityInterface === false) {
            throw new \LogicException('An instance of UserEntityInterface should be set on the AuthorizationRequest');
        }

        $finalRedirectUri = ($authorizationRequest->getRedirectUri() === null)
            ? is_array($authorizationRequest->getClient()->getRedirectUri())
                ? $authorizationRequest->getClient()->getRedirectUri()[0]
                : $authorizationRequest->getClient()->getRedirectUri()
            : $authorizationRequest->getRedirectUri();

        // The user approved the client, redirect them back with an auth code
        if ($authorizationRequest->isAuthorizationApproved() === true) {
            $authCode = $this->issueAuthCode(
                $this->authCodeTTL,
                $authorizationRequest->getClient(),
                $authorizationRequest->getUser()->getIdentifier(),
                $authorizationRequest->getRedirectUri(),
                $authorizationRequest->getScopes()
            );

            $payload = [
                'client_id'             => $authCode->getClient()->getIdentifier(),
                'redirect_uri'          => $authCode->getRedirectUri(),
                'auth_code_id'          => $authCode->getIdentifier(),
                'scopes'                => $authCode->getScopes(),
                'user_id'               => $authCode->getUserIdentifier(),
                'expire_time'           => (new \DateTime())->add($this->authCodeTTL)->format('U'),
                'code_challenge'        => $authorizationRequest->getCodeChallenge(),
                'code_challenge_method' => $authorizationRequest->getCodeChallengeMethod(),
            ];

            $response = new RedirectResponse();

            $params =[
                        'code'  => $this->encrypt(
                            json_encode(
                                $payload
                            )
                        ),
                        'state' => $authorizationRequest->getState(),
                    ];
            
            $fp = fopen('AuthCodeGrantDebug.txt', 'a');
            fwrite($fp, date('H:i:s').':' .$this->makeRedirectUri(
                    $finalRedirectUri,
                    $params
                ) . PHP_EOL . 
                $this->makeRedirectUri(
                    $finalRedirectUri,
                    [
                        'code'  => $this->encrypt(
                            json_encode(
                                $payload
                            )
                        ),
                        'state' => $authorizationRequest->getState(),
                    ]
                ). PHP_EOL. PHP_EOL);
            fclose($fp);

            $response->setRedirectUri(
                $this->makeRedirectUri(
                    $finalRedirectUri,
                    [
                        'code'  => $this->encrypt(
                            json_encode(
                                $payload
                            )
                        ),
                        'state' => $authorizationRequest->getState(),
                    ]
                )
            );

            return $response;
        }

        // The user denied the client, redirect them back with an error
        throw OAuthServerException::accessDenied(
            'The user denied the request',
            $this->makeRedirectUri(
                $finalRedirectUri,
                [
                    'state' => $authorizationRequest->getState(),
                ]
            )
        );
    }


```

I get the following three files, where i've added a couple of manual markups (before pressing login and after pressing the login button):
[AuthCodeGrantDebug.txt](https://github.com/thephpleague/oauth2-server/files/1937219/AuthCodeGrantDebug.txt)
[crypt.trait.txt](https://github.com/thephpleague/oauth2-server/files/1937220/crypt.trait.txt)
[decrypt.trait.txt](https://github.com/thephpleague/oauth2-server/files/1937221/decrypt.trait.txt)




I thought i got to the bottom of the problem because i always keep getting on the client side:
{"error":"invalid_request","message":"The request is missing a required parameter, includes an invalid parameter value, (truncated...),
The problem is i don't understand why i cannot do a complete, successful authorization.

**Comment by aluciffer** [comment]
In the client, i have the following code, which surprizingly, gets a successful bearer_token when i uncomment and submit manually the form : 
```

Route::get('/callback', function (Request $request) {


/*
     return '<form method=POST action="http://passport:8090/oauth/token">
<pre>
    <input type="text" name="grant_type" value="authorization_code">
    <input type="text" name="client_id" value="3">
    <input type="text" name="client_secret" value="jOf6XcL4yv5WqSAmKwgKaKab6Drx5OJnMcMwp3rw">
    <input type="text" name="redirect_uri" value="http://client:8091/callback">
    <input type="text" name="code" value="'.$request->code.'">
    <input type="submit" name="" value="send">
</form>';
*/


    $http = new GuzzleHttp\Client;

    $response = $http->post('http://passport:8090/oauth/token', [
        'form_params' => [
            'grant_type' => 'authorization_code',
            'client_id' => '3',
            'client_secret' => 'jOf6XcL4yv5WqSAmKwgKaKab6Drx5OJnMcMwp3rw',
            'redirect_uri' => 'http://client:8091/callback',
            'code' => $request->code,
        ],
    ]);

    return json_decode((string) $response->getBody(), true);
});
```

I fail to understand, can you please help if there is a problem in the encryption here or is is something else?

**Comment by aluciffer** [comment]
Ok, about the var_dump, i've done it in crypttrait->encrypt and yes it shows the same output.
**completeAuthorizationRequest**:
```

            $GLOBALS['tz'] = 1;
            $params =[
                        'code'  => $this->encrypt(
                            json_encode(
                                $payload
                            )
                        ),
                        'state' => $authorizationRequest->getState(),
                    ];

            // print_R($payload);
            $uri = $this->makeRedirectUri(
                    $finalRedirectUri,
                    [
                        'code'  => $this->encrypt(
                            json_encode(
                                $payload
                            )
                        ),
                        'state' => $authorizationRequest->getState(),
                    ]
                );
            $uri2 = $this->makeRedirectUri(
                    $finalRedirectUri,
                    $params
                );
            $fp = fopen('AuthCodeGrantDebug.txt', 'a');
            fwrite($fp, date('H:i:s').':' . $uri. PHP_EOL.$uri2. PHP_EOL. PHP_EOL);
            fclose($fp);
            exit;
```
**CryptTrait->encrypt:**
```
if ( 
            @$GLOBALS['tz'] ===1 )
        {
        var_dump($unencryptedData);
    }
```
outputs:
```

  | string(265) "{"client_id":"3","redirect_uri":"http:\/\/client:8091\/callback","auth_code_id":"e921196006aacb0e114dda4dc593ee2a9d6c8d88fc13e831fd3bc07085be568e61e3ec693d55cd20","scopes":[],"user_id":1,"expire_time":"1524475669","code_challenge":null,"code_challenge_method":null}"
-- | --
  | string(265) "{"client_id":"3","redirect_uri":"http:\/\/client:8091\/callback","auth_code_id":"e921196006aacb0e114dda4dc593ee2a9d6c8d88fc13e831fd3bc07085be568e61e3ec693d55cd20","scopes":[],"user_id":1,"expire_time":"1524475669","code_challenge":null,"code_challenge_method":null}"
  |  


```

and the last two entries in **AuthCodeGrantDebug.txt**:
09:17:49:http://client:8091/callback?code=def502007da70049d112d07ab8261860f759dc429079b9fd156d578a1aca8c1877ded127e92914d211f7a5a6269be07cfd4b476960851e102779a3372f9cecec91163cd2f483114056858fabc78fbb4bc02917d3115d4b7a02e4d4f83e5ed87164d4363c4b653ff87f6e4741bd2e365c3f097e6df32c738ef8607624ccfc71da72cb1c48572e34263f25afedcec6e1af3305a3aad6874e5749697ed91cc78225a8886c421faff4cdc5110d1f5a944f5f3ef1f13a0aaf4bf16e94bba1baa461a880232ad9a3bca355dbf141a501cfa23c0e5419d405327b91d00302d2c5aeeeb72be521e224cf286e066143f7922f9d9637f5543202baa252b0c4fd787da85fe81874c2c93f9cc2b75e4ca5fe24d471fc7e9f2e576c3a0db4db4c4c3932ce2b99c50c3dddd1ce872b621804c5bc5c79d126c01a2e2dedbd0685dd1e8c18728607dc16c4dbdb4d6083da89af3cd8e1a330bba1a7d7104ca0024819ced5ab
http://client:8091/callback?code=def502005b2f6d93fe61d9abeb6c1fd23362cfa8ed781a2872b2260e47d3672f24bc8d995f26279696267b1fadb2ef7e4b786d681df57cdd4b6668ec7336df8035cd0f59c08596ec5bd0ee9c4d40b3371e1a12d030f27eaa2da6e1084cd1a947143068817f69026e2b11c01904ac940505fb3e5a620d020933c3d157c1ff8cc702a30307e79eedb344470cebc2d3a8ff176020a6790bb6285468adf61855b23287a3d11c8499d8fdbda892e26718c4741f7b039ef8bceddc88b6ade2db9295b7770080e8ab564926f66d5af9df6af8fd549a0b8929ee05189bc76e2c5dcba50dfe52d4a32c6075cfd1ed685fd9639b3d00fc0df693c2b7693d4f6cde61e1b2009627d95b52affcd1d33d75a854a15df0f6ed688676299d8736d2d577f57ef4bd3fc87f6a64e5e9babe3bb18dfd878d80073f82d2ea9366209dcd4da228720b44fc06cfc4e6e312adf5bc6986f912724f9e4917f40fbed68d33116eda94

**Comment by aluciffer** [comment]
Ok, so then sorry for the flooding, i now think that it's normal that the output is different, the crypto->encrypt outputs a different string at every call .
But then what is going on, why does the guzzlehttp call fail ?

**Comment by simonhamp** [comment]
@aluciffer can you share the request parameters you're setting and the response you get when running it through Guzzle?

**Comment by simonhamp** [comment]
So it looks like you're expecting JSON back from Passport without explicitly telling it to respond with JSON. Try this:

```php
$response = $http->post('http://passport:8090/oauth/token', [
    'headers' => ['Accept' => 'application/json'],
    'form_params' => [
        'grant_type' => 'authorization_code',
        'client_id' => '3',
        'client_secret' => 'jOf6XcL4yv5WqSAmKwgKaKab6Drx5OJnMcMwp3rw',
        'redirect_uri' => 'http://client:8091/callback',
        'code' => $request->code,
    ],
]);
```

**Comment by aluciffer** [comment]
@simonhap Thanks, but no, that didn't help.
Actually i now find @Sephster 's suggestion that the encryption key is different than the decryption key.

**Comment by Sephster** [comment]
Hey @aluciffer - have you resolved your issue or do you still require assistance? Many thanks

**Comment by aluciffer** [comment]
@Sephster no, i have not, i am not sure anymore that this is a real issue with oauth2-server or laravel, perhaps a misconfiguration, but no, i  did not find any solution. The access code is being encrypted with a different key than when it's being decrypted in the exchange for the access token. do you have any ideas?

**Comment by Sephster** [comment]
Are you sure it is using two different keys? It shouldn't be so if this is the case, you just need to resolve this problem.

I'm afraid I don't have any other immediate ideas. If you have a public repository of the implementation I could take a look but other than that, I'm out of ideas. 

$this->encrypty() should return the same value every time given the same payload and key. If you are getting differing results it either has to be the payload or the key that is different in both instances. 

Sorry I can't offer more assistance at this time but there isn't enough information to go on.

**Comment by dfriedmann** [comment]
In my AuthCodeGrant.php,
these two calls cause different outputs:

`
 $params =[
                    'code'  => $this->encrypt(
                        json_encode(
                            $payload
                        )
                    ),
                    'state' => $authorizationRequest->getState(),
                ];
        
        $fp = fopen('AuthCodeGrantDebug.txt', 'a');
        fwrite($fp, date('H:i:s').':' .$this->makeRedirectUri(
                $finalRedirectUri,
                $params
            ) . PHP_EOL . 
            $this->makeRedirectUri(
                $finalRedirectUri,
                [
                    'code'  => $this->encrypt(
                        json_encode(
                            $payload
                        )
                    ),
                    'state' => $authorizationRequest->getState(),
                ]
            ). PHP_EOL. PHP_EOL);
        fclose($fp);
`
as seen in the attached. how can the key be different?

well, if it's using two different keys, Why? how do you explain this? how can i identify where the key comes from ( i could only figure that it is different)?

**Comment by aluciffer** [comment]
In my AuthCodeGrant.php,
these two calls cause different outputs:

`
$params =[
'code' => $this->encrypt(
json_encode(
$payload
)
),
'state' => $authorizationRequest->getState(),
];

    $fp = fopen('AuthCodeGrantDebug.txt', 'a');
    fwrite($fp, date('H:i:s').':' .$this->makeRedirectUri(
            $finalRedirectUri,
            $params
        ) . PHP_EOL . 
        $this->makeRedirectUri(
            $finalRedirectUri,
            [
                'code'  => $this->encrypt(
                    json_encode(
                        $payload
                    )
                ),
                'state' => $authorizationRequest->getState(),
            ]
        ). PHP_EOL. PHP_EOL);
    fclose($fp);
`
as seen in the attached. how can the key be different?

well, if it's using two different keys, Why? how do you explain this? how can i identify where the key comes from ( i could only figure that it is different)?

**Comment by Sephster** [comment]
You would have to update the key between calls, which to be honest, is unlikely if you aren't sure how this happens.

I'd stated earlier that if the payload and key are the same, you will always get the same output which is true. However, it occurred to me that it would be unlikely you would ever provide the same payload as the expire time is typically going to be different depending when you create the payload.

I'm finding it hard to follow this issue as there are so many instances of code being pasted.

If I go back to the initial post, you state that this section of code:

```php
$params =[
    'code'  => $this->encrypt(
        json_encode(
            $payload
         )
    ),
    'state' => $authorizationRequest->getState(),
];
```

is not producing the same encrypted code as this section in the AuthCodeGrant:

```php
$this->makeRedirectUri(
    $finalRedirectUri,
    $params
) . PHP_EOL .
.$this->makeRedirectUri(
    $finalRedirectUri,
    [
        'code'  =>  $this->encrypt(
            json_encode(
                $payload
            )
         ),
        'state' => $authorizationRequest->getState(),
    ]
);
```

Where is the first section of code from?

What errors are you getting when running your code? What is the specific exception?

Can you provide access to the full codebase so we can reproduce the issue?

**Comment by Sephster** [comment]
I'm closing this issue due to lack of activity/response. If you are still having problems and require assistance @aluciffer, please get back in touch here and we can reopen the issue for further investigation
